### ISSUE TITLE
Stream(...., reset: true) not working on child component

### ISSUE THREAD

**Issue by Ali-Doggaz** [phoenixframework/phoenix_live_view#3136]
### Environment

* Elixir version (elixir -v): 1.16.1
* Phoenix version (mix deps):1.7.10
* Phoenix LiveView version (mix deps):0.20.9
* Operating system: Mac OS
* Browsers you attempted to reproduce this bug on (the more the merrier): Brave
* Does the problem persist after removing "assets/node_modules" and trying again? Yes/no: no

### Actual behavior
The child component used to render a table from a streamed list does not remove rows that have been deleted from the stream after calling stream(...., reset: true). 

More context:

Here is the table component: 
```elixir
    ~H"""
    <table class=...>
      <thead class=..>
        ...
      </thead>
      <tbody
        id={@id}
        phx-update={match?(%Phoenix.LiveView.LiveStream{}, @rows) && "stream"}
        class="..."
      >
        <tr
          :for={row <- @rows}
          id={@row_id && @row_id.(row)}
          class="..."
        >
          <td
            :for={{col, i} <- Enum.with_index(@col)}
            phx-click={@row_click && @row_click.(row)}
            class=...
          >
            <div class=...>
              <%= render_slot(col, @row_item.(row)) %>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    """
```

```elixir
  <MyComponents.table id="items_table" class="w-full mb-5" rows={@streams.items}>
    <:col :let={{_id, item}} label="...">
        <%= item.inserted_at %>
      </span>
    </:col>
...
  </MyComponents.table>
```

When the "fitlers_changed" event is triggered, I have a handle_event function that resets the items:
```elixir
 def handle_event("filter_changed", params, socket) do
    filters = get_filters(params)
    new_items = get_new_items(@page_size_limit, filters)
    
   # Update the URL's query params
    updated_url = build_full_url(filters)

    # Update the assigns
    {:noreply,
     socket
     |> assign(:filters, filters)
     |> stream(:items, new_items, reset: true)
     |> push_patch(to: updated_url)}
  end
```

=> After changing the filters, new items are added to the table but the previous items which were supposed to get deleted by `reset:true` are still there. 
Instead of using a child component, I tried adding this to my parent's view html:
```elixir
  <div>
    <ul>
      <%= for {id, item} <- @streams.items do %>
        <li><%= item.id %></li>
      <% end %>
    </ul>
  </div>
```
And it worked perfectly (old rows were getting deleted properly).

**Comment by SteffenDE** [comment]
Hi, thank you for reporting this issue. I'll be happy to look into it. Can you please create a minimal example to reproduce using the following template? You can save this as `myfile.exs` and execute it with `elixir myfile.exs`:

```elixir
Application.put_env(:sample, Example.Endpoint,
  http: [ip: {127, 0, 0, 1}, port: 5001],
  server: true,
  live_view: [signing_salt: "aaaaaaaa"],
  secret_key_base: String.duplicate("a", 64)
)

Mix.install([
  {:plug_cowboy, "~> 2.5"},
  {:jason, "~> 1.0"},
  {:phoenix, "~> 1.7"},
  # please test your issue using the latest version of LV from GitHub!
  {:phoenix_live_view, github: "phoenixframework/phoenix_live_view", branch: "main", override: true},
])

# build the LiveView JavaScript assets (this needs mix and npm available in your path!)
path = Phoenix.LiveView.__info__(:compile)[:source] |> Path.dirname() |> Path.join("../")
System.cmd("mix", ["deps.get"], cd: path, into: IO.binstream())
System.cmd("npm", ["install"], cd: Path.join(path, "./assets"), into: IO.binstream())
System.cmd("mix", ["assets.build"], cd: path, into: IO.binstream())

defmodule Example.ErrorView do
  def render(template, _), do: Phoenix.Controller.status_message_from_template(template)
end

defmodule Example.HomeLive do
  use Phoenix.LiveView, layout: {__MODULE__, :live}

  def mount(_params, _session, socket) do
    {:ok, assign(socket, :count, 0)}
  end

  def render("live.html", assigns) do
    ~H"""
    <script src="/assets/phoenix/phoenix.js"></script>
    <script src="/assets/phoenix_live_view/phoenix_live_view.js"></script>
    <%!-- uncomment to use enable tailwind --%>
    <%!-- <script src="https://cdn.tailwindcss.com"></script> --%>
    <script>
      let liveSocket = new window.LiveView.LiveSocket("/live", window.Phoenix.Socket)
      liveSocket.connect()
    </script>
    <style>
      * { font-size: 1.1em; }
    </style>
    <%= @inner_content %>
    """
  end

  def render(assigns) do
    ~H"""
    <%= @count %>
    <button phx-click="inc">+</button>
    <button phx-click="dec">-</button>
    """
  end

  def handle_event("inc", _params, socket) do
    {:noreply, assign(socket, :count, socket.assigns.count + 1)}
  end

  def handle_event("dec", _params, socket) do
    {:noreply, assign(socket, :count, socket.assigns.count - 1)}
  end
end

defmodule Example.Router do
  use Phoenix.Router
  import Phoenix.LiveView.Router

  pipeline :browser do
    plug(:accepts, ["html"])
  end

  scope "/", Example do
    pipe_through(:browser)

    live("/", HomeLive, :index)
  end
end

defmodule Example.Endpoint do
  use Phoenix.Endpoint, otp_app: :sample
  socket("/live", Phoenix.LiveView.Socket)

  plug Plug.Static, from: {:phoenix, "priv/static"}, at: "/assets/phoenix"
  plug Plug.Static, from: {:phoenix_live_view, "priv/static"}, at: "/assets/phoenix_live_view"

  plug(Example.Router)
end

{:ok, _} = Supervisor.start_link([Example.Endpoint], strategy: :one_for_one)
Process.sleep(:infinity)
```

This would help tremendously.

**Comment by Ali-Doggaz** [comment]
Thank you for your quick answer @SteffenDE !

I tried replicating the problem using your template with no luck... BUT my attempts led me to discover a couple of interesting things about this problem:

1. The problem is not about the child/parent relationship. I added this div directly inside my parent's view and the problem persisted:
```elixir
  <table class="divide-y divide-gray-200 border-t border-gray-200 text-sm dark:divide-slate-700 dark:border-slate-600">
    <tbody id="test_table_3" phx-update="stream">
      <tr :for={{id, billing_item} <- @streams.billing_items} id={id}>
        <td>
          <%= billing_item.id %>
        </td>
      </tr>
    </tbody>
  </table>
```

2. Now the most interesting part: I discovered that removing the `phx-update="stream"` fixes the problem. Did I miss something? Initially, I added this attribute because I read in the docs:
`Note: Failing to place phx-update="stream" on the immediate parent for each stream will result in broken behavior.` 
3. After trying to use your template to replicate the problem and creating a simple stream, I noticed that "normal" stream elements get a `data-phx-stream` attribute. When I inspected my real project's html, this attribute was missing on all my stream items rows.

I still don't know what is causing this problem and cannot replicate it, but I will try again next week. In the meantime, if someone has an idea about how to replicate the bug here is my "template":
```elixir
Application.put_env(:sample, Example.Endpoint,
  http: [ip: {127, 0, 0, 1}, port: 5001],
  server: true,
  live_view: [signing_salt: "aaaaaaaa"],
  secret_key_base: String.duplicate("a", 64)
)

Mix.install([
  {:plug_cowboy, "~> 2.5"},
  {:jason, "~> 1.0"},
  {:phoenix, "~> 1.7"},
  # please test your issue using the latest version of LV from GitHub!
  {:phoenix_live_view, github: "phoenixframework/phoenix_live_view", branch: "main", override: true},
])

# build the LiveView JavaScript assets (this needs mix and npm available in your path!)
path = Phoenix.LiveView.__info__(:compile)[:source] |> Path.dirname() |> Path.join("../")
System.cmd("mix", ["deps.get"], cd: path, into: IO.binstream())
System.cmd("npm", ["install"], cd: Path.join(path, "./assets"), into: IO.binstream())
System.cmd("mix", ["assets.build"], cd: path, into: IO.binstream())

defmodule Example.ErrorView do
  def render(template, _), do: Phoenix.Controller.status_message_from_template(template)
end

defmodule Example.HomeLive do
  use Phoenix.LiveView, layout: {__MODULE__, :live}

  def mount(_params, _session, socket) do
    items = [%{id: 1, value: "one"}, %{id: 2, value: "two"}]


    {:ok, assign(socket, :count, 0)
    |> stream(:items, items)}
  end

  def handle_params(_params, _session, socket) do
    {:noreply, socket}
  end

  def render("live.html", assigns) do
    ~H"""
    <script src="/assets/phoenix/phoenix.js"></script>
    <script src="/assets/phoenix_live_view/phoenix_live_view.js"></script>
    <%!-- uncomment to use enable tailwind --%>
    <%!-- <script src="https://cdn.tailwindcss.com"></script> --%>
    <script>
      let liveSocket = new window.LiveView.LiveSocket("/live", window.Phoenix.Socket)
      liveSocket.connect()
    </script>
    <style>
      * { font-size: 1.1em; }
    </style>
    <%= @inner_content %>
    """
  end



  def render(assigns) do
    ~H"""
    <table>
      <tbody phx-update="stream" id="test_table_3">
        <tr :for={{id, item} <- @streams.items} id={id}>
          <td >
            <%= item.id %>
          </td>
        </tr>
      </tbody>
    </table>

    <button phx-click="filter">Filter</button>

    """
  end


  def working_render(assigns) do
    ~H"""
    <ul>
      <%= for {id, item }<- @streams.items do %>
        <li><%= item.value %></li>
      <% end %>
    </ul>
    <%= @count %>
    <button phx-click="filter">Filter</button>
    """
  end

  def handle_event("filter", _params, socket) do
    new_items = [%{id: 3, value: "Three"}]
    {:noreply, socket |> stream(:items, new_items, reset: true) |> push_patch(to: "/")}
  end
end

defmodule Example.Router do
  use Phoenix.Router
  import Phoenix.LiveView.Router

  pipeline :browser do
    plug(:accepts, ["html"])
  end

  scope "/", Example do
    pipe_through(:browser)

    live("/", HomeLive, :index)
  end
end

defmodule Example.Endpoint do
  use Phoenix.Endpoint, otp_app: :sample
  socket("/live", Phoenix.LiveView.Socket)

  plug Plug.Static, from: {:phoenix, "priv/static"}, at: "/assets/phoenix"
  plug Plug.Static, from: {:phoenix_live_view, "priv/static"}, at: "/assets/phoenix_live_view"

  plug(Example.Router)
end

{:ok, _} = Supervisor.start_link([Example.Endpoint], strategy: :one_for_one)
Process.sleep(:infinity)

```

**Comment by Oliphaunte** [comment]
Apologies ahead of time if I steer this down the wrong path.

I recently had this problem too, I believe it is related to the table component itself, reset:true did not work at all for me using it.

I resolved it by adding the line `row_id={fn {dom_id, _} -> dom_id end}` to the table component.

```
   <.table id="items_table" class="w-full mb-5" rows={@streams.items} row_id={fn {dom_id, _} -> dom_id end}>
      <:col :let={{_id, item}} label="...">
          <%= item.value %>
      </:col>
    </.table>
 ```

The table component does have `assign(assigns, row_id: assigns.row_id || fn {id, _item} -> id end)` which should be doing the same thing, so not sure if maybe there is a bug there?
 
 Here the full input I used to test it out. If you remove the `row_id` line, the stream is no longer reset correctly. 
 
```
 Application.put_env(:sample, Example.Endpoint,
  http: [ip: {127, 0, 0, 1}, port: 5001],
  server: true,
  live_view: [signing_salt: "aaaaaaaa"],
  secret_key_base: String.duplicate("a", 64)
)

Mix.install([
  {:plug_cowboy, "~> 2.5"},
  {:phoenix_ecto, "~> 4.4"},
  {:phoenix_html, "~> 4.0", override: true},
  # TODO (ts): Remove this dependency once the `use` clause is not necessary anymore (`use Phoenix.HTML` is replaced by `use PhoenixHTMLHelpers` because the former is deprecated).
  {:phoenix_html_helpers, "~> 1.0"},
  {:phoenix_live_dashboard, "~> 0.8.0"},
  {:phoenix_live_reload, "~> 1.3", only: :dev},
  {:phoenix_live_view,
   github: "phoenixframework/phoenix_live_view", branch: "main", override: true},
  {:phoenix_pubsub, "~> 2.1.1"},
  {:phoenix_swoosh, "~> 1.0"},
  {:jason, "~> 1.3"},
  {:phoenix, "~> 1.7.0-rc.2", override: true},
  {:plug, "~> 1.14"}
])

# build the LiveView JavaScript assets (this needs mix and npm available in your path!)
path = Phoenix.LiveView.__info__(:compile)[:source] |> Path.dirname() |> Path.join("../")
System.cmd("mix", ["deps.get"], cd: path, into: IO.binstream())
System.cmd("npm", ["install"], cd: Path.join(path, "./assets"), into: IO.binstream())
System.cmd("mix", ["assets.build"], cd: path, into: IO.binstream())

defmodule Example.ErrorView do
  def render(template, _), do: Phoenix.Controller.status_message_from_template(template)
end

defmodule Example.HomeLive do
  use Phoenix.LiveView, layout: {__MODULE__, :live}

  def mount(_params, _session, socket) do
    items = [%{id: 1, value: "one"}, %{id: 2, value: "two"}]

    {:ok,
     assign(socket, :count, 0)
     |> stream(:items, items)}
  end

  def handle_params(_params, _session, socket) do
    {:noreply, socket}
  end

  def render("live.html", assigns) do
    ~H"""
    <script src="/assets/phoenix/phoenix.js"></script>
    <script src="/assets/phoenix_live_view/phoenix_live_view.js"></script>
    <%!-- uncomment to use enable tailwind --%>
    <%!-- <script src="https://cdn.tailwindcss.com"></script> --%>
    <script>
      let liveSocket = new window.LiveView.LiveSocket("/live", window.Phoenix.Socket)
      liveSocket.connect()
    </script>
    <style>
      * { font-size: 1.1em; }
    </style>
    <%= @inner_content %>
    """
  end

  def render(assigns) do
    ~H"""
    <.table id="items_table" class="w-full mb-5" rows={@streams.items} row_id={fn {dom_id, _} -> dom_id end}>
      <:col :let={{_id, item}} label="...">
          <%= item.value %>
      </:col>
    </.table>

    <button phx-click="filter">Filter</button>

    """
  end

  @doc ~S"""
  Renders a table with generic styling.

  ## Examples

      <.table id="users" rows={@users}>
        <:col :let={user} label="id"><%= user.id %></:col>
        <:col :let={user} label="username"><%= user.username %></:col>
      </.table>
  """
  attr(:id, :string, required: true)
  attr(:rows, :list, required: true)
  attr(:row_id, :any, default: nil, doc: "the function for generating the row id")
  attr(:row_click, :any, default: nil, doc: "the function for handling phx-click on each row")
  attr(:row_class, :string, default: nil)

  attr(:row_item, :any,
    default: &Function.identity/1,
    doc: "the function for mapping each row before calling the :col and :action slots"
  )

  slot :col, required: true do
    attr(:label, :string)
    attr(:class, :string)
  end

  slot(:action, doc: "the slot for showing user actions in the last table column")

  def table(assigns) do
    assigns =
      with %{rows: %Phoenix.LiveView.LiveStream{}} <- assigns do
        assign(assigns, row_id: assigns.row_id || fn {id, _item} -> id end)
        assign(assigns, row_class: assigns.row_class || fn {_, _} -> "group hover:bg-zinc-50" end)
      end

    ~H"""
    <div class="overflow-y-auto px-4 sm:overflow-visible sm:px-0">
      <table class="w-[40rem] sm:w-full">
        <thead class="sticky top-0 z-10 bg-white text-sm text-left leading-6 text-zinc-500">
          <tr>
            <th :for={col <- @col} class={if Map.get(col, :class), do: col[:class], else: "p-0 py-4 pr-6 font-normal"}>
              <%= col[:label] %>
            </th>
            <th :if={@action != []} class="relative p-0 pb-4">
              <span class="sr-only"><%= "Actions" %></span>
            </th>
          </tr>
        </thead>

        <tbody
          id={@id}
          phx-update={match?(%Phoenix.LiveView.LiveStream{}, @rows) && "stream"}
          class="relative divide-y divide-zinc-100 border-t border-zinc-200 text-sm leading-6 text-zinc-700"
        >
          <tr :for={row <- @rows} id={@row_id && @row_id.(row)} class={@row_class && @row_class.(row)}>
            <td
              :for={{col, i} <- Enum.with_index(@col)}
              phx-click={@row_click && @row_click.(row)}
              class={["relative p-0", @row_click && "hover:cursor-pointer"]}
            >
              <div class="block py-4 pr-6">
                <span class="absolute -inset-y-px right-0 -left-4 group-hover:bg-zinc-50 sm:rounded-l-xl" />
                <span class={["relative", i == 0 && "font-semibold text-zinc-900"]}>
                  <%= render_slot(col, @row_item.(row)) %>
                </span>
              </div>
            </td>
            <td :if={@action != []} class="relative w-14 p-0">
              <div class="relative whitespace-nowrap py-4 text-right text-sm font-medium">
                <span class="absolute -inset-y-px -right-4 left-0 group-hover:bg-zinc-50 sm:rounded-r-xl" />
                <span :for={action <- @action} class="relative ml-4 font-semibold leading-6 text-zinc-900 hover:text-zinc-700">
                  <%= render_slot(action, @row_item.(row)) %>
                </span>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    """
  end

  def handle_event("filter", _params, socket) do
    new_items = [%{id: 3, value: "Three"}]
    {:noreply, socket |> stream(:items, new_items, reset: true) |> push_patch(to: "/")}
  end
end

defmodule Example.Router do
  use Phoenix.Router
  import Phoenix.LiveView.Router

  pipeline :browser do
    plug(:accepts, ["html"])
  end

  scope "/", Example do
    pipe_through(:browser)

    live("/", HomeLive, :index)
  end
end

defmodule Example.Endpoint do
  use Phoenix.Endpoint, otp_app: :sample
  socket("/live", Phoenix.LiveView.Socket)

  plug(Plug.Static, from: {:phoenix, "priv/static"}, at: "/assets/phoenix")
  plug(Plug.Static, from: {:phoenix_live_view, "priv/static"}, at: "/assets/phoenix_live_view")

  plug(Example.Router)
end

{:ok, _} = Supervisor.start_link([Example.Endpoint], strategy: :one_for_one)
Process.sleep(:infinity)
```

**Comment by SteffenDE** [comment]
@Oliphaunte there's a bug in your table component. The first assign statement is ignored.

```elixir
  assigns =
      with %{rows: %Phoenix.LiveView.LiveStream{}} <- assigns do
        assign(assigns, row_id: assigns.row_id || fn {id, _item} -> id end)
        assign(assigns, row_class: assigns.row_class || fn {_, _} -> "group hover:bg-zinc-50" end)
      end
```

should be

```elixir
  assigns =
      with %{rows: %Phoenix.LiveView.LiveStream{}} <- assigns do
        assigns
        |> assign(row_id: assigns.row_id || fn {id, _item} -> id end)
        |> assign(row_class: assigns.row_class || fn {_, _} -> "group hover:bg-zinc-50" end)
      end
```

**Comment by Oliphaunte** [comment]
@SteffenDE Yep, thanks, was using the wrong table component.

**Comment by Ali-Doggaz** [comment]
Alright, I think this is one of the weirdest bugs I saw in my career (which is still relatively new 👻):
After spending days on this, I had to ask one of my coworkers for help. My coworkers cloned my code, installed the exact same dependencies, and ran it.
The stream(..., reset: true) feature worked properly for him!!! When he inspected his HTML, he found the `data-phx-stream` attribute correctly added to the rows. 

We use the same OS (Mac OS), same browser (and we tested with 3 different ones: safari, brave, chrome), same code, same deps (identical mix.exs and mix.lock).

I have NO IDEA what could be causing the `data-phx-stream` attribute to not be added on my browser. No errors in the console or anything... If anyone has a guess I'd be more than happy to test it.
